### PR TITLE
Let devices register "available commands" with associated metadata.

### PR DESCRIPTION
**Pull request by rfk**
This is a revision of the "device capabilities" feature, to allow devices to associate a metadata blob with each advertised capability.  I've taken the liberty of renaming them from "capabilities" to "commands" for reasons that will become clear in a forthcoming auth-server PR.

WIP for now. /cc @eoger 

Connects to https://github.com/mozilla/fxa-auth-server/issues/2318

### CODE REVIEW

File: lib/db/schema/patch-080-081.sql [hunk 1]
```diff
@@ -0,0 +1,308 @@
+--
```
Comment (eoger) [inline lib/db/schema/patch-080-081.sql]: Should we remove the `capabilities` table/code, and in fxa-auth-server just accept `capabilities` in the incoming device bodies?

Comment (rfk) [reply lib/db/schema/patch-080-081.sql]: Some of the existing stored procedures try to pull data from the `capabilities` table, so we can't drop it until we've fully rolled out the updated stored procedures.

**Comment by rfk** [comment]
I have this up and running in a dev environment, and it seems to be working well!  I think it's ready for a preliminary r?, although I'll leave a few comments inline...

File: db-server/index.js [hunk 2]
```diff
@@ -177,6 +176,14 @@ function createServer(db) {
     }
   }
 
+  api.get('/account/:id/devices', withIdAndBody(db.accountDevices))
+
+  api.get(
+    '/account/:uid/device/:deviceId',
```
Comment (rfk) [inline db-server/index.js]: Over in the auth-server, whenever we want to get the device record for a particular device, we get it for *all* devices and then filter down to just the one we want.  I added an explicit "get the record for one specific device" method here so we can avoid that in future.

File: db-server/lib/error.js [hunk 3]
```diff
@@ -83,17 +83,6 @@ AppError.invalidVerificationMethod = function () {
...
-AppError.unknownDeviceCapability = function () {
-  return new AppError(
-    {
-      code: 400,
-      error: 'Bad request',
-      errno: 139,
-      message: 'Unknown device capability'
-    }
-  )
-}
```
Comment (rfk) [inline db-server/lib/error.js]: @eoger points out that we don't have any code depending on this errno, so it's safe to just remove it rather than deprecating it.

File: lib/db/mysql.js [hunk 4]
```diff
@@ -1066,7 +1116,7 @@ module.exports = function (log, error) {
       })
   }
 
-  MySql.prototype.readOneFromFirstResult = function (sql, params) {
+  MySql.prototype.readAllResults = function (sql, params) {
```
Comment (rfk) [inline lib/db/mysql.js]: Coming back to this code after a while, I was confused by the name "read one from first result" because as far as I could tell, it was reading multiple things.  I've renamed it to just "read all results" which seems a closer match to what it does, but I'm happy to be schooled on why that's not accurate if it isn't.

File: lib/db/mysql.js [hunk 5]
```diff
@@ -413,92 +407,148 @@ module.exports = function (log, error) {
...
   //          AS deviceCreatedAt, d.callbackURL AS deviceCallbackURL, d.callbackPublicKey
   //          AS deviceCallbackPublicKey, d.callbackAuthKey AS deviceCallbackAuthKey,
-  //          d.callbackIsExpired AS deviceCallbackIsExpired,
-  //          dc.capabilities AS deviceCapabilities
+  //          d.callbackIsExpired AS deviceCallbackIsExpired
   //          ut.tokenVerificationId, ut.mustVerify
   // Where  : t.tokenId = $1 AND t.uid = a.uid AND t.tokenId = d.sessionTokenId AND
   //          t.uid = d.uid AND t.tokenId = u.tokenId
-  var SESSION_DEVICE = 'CALL sessionWithDevice_13(?)'
+  var SESSION_DEVICE = 'CALL sessionWithDevice_14(?)'
```
Comment (rfk) [inline lib/db/mysql.js]: So, I'm slightly concerned about doing this additional aggregation work in `sessionWithDevice`.  We call `sessionWithDevice` **a lot** - it's the primary auth method for a bunch of auth-server calls, almost none of which actually need the `availableCommands` bundle.  Is it worth re-introducing a plain old "session" method that doesn't do this extra work, and keeping the more expensive version for when we actually need the full device record?

IIRC we often need *some* of the device metadata, which is why we moved everything to use `sessionWithDevice`, is that accurate?

**Comment by rfk** [comment]
See also the corresponding auth-server PR, which has more to say about how this functionality is used: https://github.com/mozilla/fxa-auth-server/pull/2449

File: db-server/index.js [hunk 6]
```diff
@@ -177,6 +176,14 @@ function createServer(db) {
...
+    op(function (req) {
+      return db.device(req.params.uid, req.params.deviceId)
```
Comment (philbooth) [inline db-server/index.js]: Not sure if it makes sense or is the correct convention, ignore if not, but can this and the called method be tweaked to use the `withParams` helper here?

File: db-server/test/backend/db_tests.js [hunk 7]
```diff
@@ -992,38 +1008,59 @@ module.exports = function (config, DB) {
...
+        const newDevice = Object.assign({}, deviceInfo, {
+          availableCommands: {}
+        })
         return db.updateDevice(accountData.uid, deviceInfo.deviceId, newDevice)
           .then(() => {
-            return db.accountDevices(accountData.uid)
+            return db.device(accountData.uid, deviceInfo.deviceId)
           })
-          .then((devices) => assert.deepEqual(devices[0].capabilities, ['messages']))
+          .then(device => assert.deepEqual(device.availableCommands, {}))
```
Comment (philbooth) [inline db-server/test/backend/db_tests.js]: Does this just duplicate the assertion on line `977`, or is there a difference between them?

File: lib/db/mem.js [hunk 8]
```diff
@@ -474,15 +463,14 @@ module.exports = function (log, error) {
         function(account) {
           return Object.keys(account.devices)
             .map(
-              function (id) {
-                var device = account.devices[id]
+              function (deviceKey) {
+                var device = Object.assign({}, account.devices[deviceKey])
```
Comment (philbooth) [inline lib/db/mem.js]: @shane-tomlinson showed me this sweet trick the other day, [object spreading](https://github.com/tc39/proposal-object-rest-spread#spread-properties):

```js
var device = { ...account.devices[deviceKey] }
```

Not that we have to use it of course, but it's my brand new hammer and all I see are nails.

File: lib/db/mysql.js [hunk 9]
```diff
@@ -308,79 +308,73 @@ module.exports = function (log, error) {
     )
   }
 
-  const ADD_CAPABILITY = 'CALL addCapability_1(?, ?, ?)'
-  const PURGE_CAPABILITIES = 'CALL purgeCapabilities_1(?, ?)'
+  const UPSERT_AVAILABLE_COMMAND = 'CALL upsertAvailableCommand_1(?, ?, ?, ?)'
+  const PURGE_AVAILABLE_COMMANDS = 'CALL purgeAvailableCommands_1(?, ?)'
 
   var CREATE_DEVICE = 'CALL createDevice_4(?, ?, ?, ?, ?, ?, ?, ?, ?)'
```
Comment (philbooth) [inline lib/db/mysql.js]: I didn't spot this when it happened before, but can you move this definition of `CREATE_DEVICE` down so it's next to the `createDevice` method again?

File: lib/db/mysql.js [hunk 10]
```diff
@@ -413,92 +407,148 @@ module.exports = function (log, error) {
...
+  //    { id: 'three', other: 'baz', agg: {}},
+  //  ]
+  //
+  // Note that correct operation depends on the rows being ordered by id column,
+  // so they can be aggregated in a single pass.
+  //
+  // The column names can be customized, in order to work with several different
+  // queries.
+
+  function aggregateNameValuePairs(rows, idColumn, nameColumn, valueColumn, aggColumn) {
```
Comment (philbooth) [inline lib/db/mysql.js]: Is this something that warrants its own module with some dedicated unit tests?

Comment (philbooth) [reply lib/db/mysql.js]: (the unit tests is really the bit I'm asking about here tbh, regardless of the separate module part)

File: lib/db/mysql.js [hunk 11]
```diff
@@ -413,92 +407,148 @@ module.exports = function (log, error) {
...
-    return result
+      if (row[nameColumn]) {
+        curItem[aggColumn][row[nameColumn]] = row[valueColumn]
+      }
+      return items
+    }, [])
   }
 
-  // Select : devices d, deviceCapabilities dc, sessionTokens s, accounts a
+  // Select : devices d, sessionTokens s, deviceAvailableCommands sc, deviceCommandIdentifiers ci
```
Comment (philbooth) [inline lib/db/mysql.js]: Just out of curiosity, what does the "s" stand for in `sc`? (the stored procedure still has it as `dc` I notice, not that it matters)

Comment (rfk) [reply lib/db/mysql.js]: Nice catch; these used to be "supported commands" rather than "available commands" so I guess it's leftover from that naming convention.

File: lib/db/mysql.js [hunk 12]
```diff
@@ -413,92 +407,148 @@ module.exports = function (log, error) {
...
+  //    { id: 'one', other: 'foo', agg: { 'name1': 'value1' }},
+  //    { id: 'two', other: 'bar', agg: { 'name1': 'value1', 'name2': 'value1' }},
+  //    { id: 'three', other: 'baz', agg: {}},
+  //  ]
+  //
+  // Note that correct operation depends on the rows being ordered by id column,
+  // so they can be aggregated in a single pass.
+  //
+  // The column names can be customized, in order to work with several different
+  // queries.
```
Comment (philbooth) [inline lib/db/mysql.js]: Even with this and the inline commentary, I'm finding it hard to follow what this function does. Few suggestions, feel free to ignore:

* If this comment was more docblocky, with definitions for each argument etc, I think I'd find it easier to understand. For instance, what is `aggColumn`? From reading the code, I think it's just the name of the property that is mapped to `agg` in your example output above, but it would be good to have those definitions laid out formally.

* Alternatively, just add the example function call that produces the output above. We have the input definition and the output definition, so maybe just seeing what the function call is with hard-coded values for the `...Column` params might be enough.

* It might be easier to follow if the example used real-life column names, keys and values from the db schema. The fact that the example data is so abstract makes it hard to pin down in my head.

* It's poor reading comprehension on my part, but it took me ages to spot that `id` is not unique in the example input data (only just spotted it while I was typing the last bullet). That's the key transformation here right, we're normalizing the data so that each `id` is unique in the result set? Might be worth calling that out more explicitly, if I understand it correctly.

Comment (philbooth) [reply lib/db/mysql.js]: (and if the last point is on the right track, I wonder whether calling the function `normalizeNameValuePairs` might help too? or is the word "normalize" just imbued with unwarranted significance in my head?)

File: lib/db/mysql.js (hunk 5 above)
Comment (philbooth) [reply lib/db/mysql.js]: How many `availableCommands` do we reasonably expect in practice? It's worth measuring of course but in the context of a joined query against our `sessionTokens` and `devices` tables, it looks like the time spent inside `aggregateNameValuePairs` might be a drop in the ocean to me.

Comment (philbooth) [reply lib/db/mysql.js]: > IIRC we often need *some* of the device metadata, which is why we moved everything to use `sessionWithDevice`, is that accurate?

Yep, I think so. Actually, it's good you asked this question, I think I just found a bug while looking in to it.

File: lib/db/schema/patch-081-080.sql [hunk 13]
```diff
@@ -0,0 +1,17 @@
+-- -- TODO: Implement your *commented-out* reverse migration here
```
Comment (philbooth) [inline lib/db/schema/patch-081-080.sql]: We probably don't want this left in!

File: lib/db/util.js [hunk 14]
```diff
@@ -41,26 +41,8 @@ const VERIFICATION_METHODS = new Map([
...
 module.exports = {
 
-  mapDeviceCapability(val) {
-    if (typeof val === 'number') {
-      return DEVICE_CAPABILITIES_IDS.get(val) || null
-    } else {
-      return DEVICE_CAPABILITIES.get(val) || null
-    }
-  },
-
```
Comment (philbooth) [inline lib/db/util.js]: :+1:

File: lib/db/mysql.js (hunk 5 above)
Comment (rfk) [reply lib/db/mysql.js]: Probably not that many in practice, so perhaps we can take a wait-and-see approach here and optimize it more holistically if it turns out to be a growing bottleneck in practice.

File: db-server/test/backend/db_tests.js (hunk 7 above)
Comment (rfk) [reply db-server/test/backend/db_tests.js]: You're right, it's a dupe, I'll remove it.

File: db-server/index.js (hunk 6 above)
Comment (rfk) [reply db-server/index.js]: I tried this out, but it made the tests uglier because you have to call e.g. `db.createDevice({uid: account.uid, deviceId: deviceinfo.id }, deviceInfo)`.  I'm adding a modified helper that I like better, but YMMV, I'd be interested to hear what you think.

**Comment by rfk** [comment]
@philbooth I've pushed my review responses, let me know if they all seem OK.  I'll come back to this later on to rebase against master for final merge, given conflicting db migration.

File: db-server/index.js [hunk 15]
```diff
@@ -41,12 +41,24 @@ function createServer(db) {
...
     return reply(function (params, body, query) {
       return fn.call(db, params, body)
     })
   }
 
+  function withSpreadParamsAndBody(fn) {
+    return reply(function (params, body, query) {
+      return fn.apply(db, Object.keys(params).map(k => params[k]).concat([body]))
+    })
+  }
```
Comment (philbooth) [inline db-server/index.js]: Yep, I like these a lot! :+1:

File: lib/db/util.js [hunk 16]
```diff
@@ -134,5 +116,84 @@ module.exports = {
...
+  //    ")
+  //
+  // Then apply `aggregateNameValuePairs` on the rows, providing the names of the
+  // columns to use for the group-by id, the keys, the values, and the resulting
+  // aggregate:
+  //
+  //    unique_rows = aggregateNameValuePairs(rows, "id", "name", "value", "result")
+  //
+  // The end result will be the same as produced by JSON_OBJECTAGG, including correct
+  // handling of NULL values in all columns.
```
Comment (philbooth) [inline lib/db/util.js]: Perfect, thanks!

File: lib/db/mem.js (hunk 8 above)
Comment (rfk) [reply lib/db/mem.js]: I like this, but unfortunately `grunt eslint` seems to choke on it, and I don't have the bandwidth to debug right now so I'm going to revert.
